doc/source/numpy_2_0_migration_guide.rst

```diff
@@ -0,0 +1,147 @@
+*************************
+NumPy 2.0 Migration Guide
+*************************
+
+This document contains a set of instructions on how to update your codebase to NumPy 2.0
+Python API compatible version. Most of the changes are trivial, and require the end user
+to use a different name/module to access a given function/constant.
+
+Please refer to `NEP 52 <https://numpy.org/neps/nep-0052-python-api-cleanup.html>`_ for more details.
+
+
+Main namespace
+--------------
+
+About 100 members of the main ``np`` namespace has been deprecated, removed or 
+moved to a new place. It was done to reduce clutter and establish only one way to
+access a given attribute. The table below shows members that have been removed:
+
+======================  =================================================================
+removed member          migration guidline
+======================  =================================================================
+add_docstring           It's still available as ``np.lib.add_docstring``.
+add_newdoc              It's still available as ``np.lib.add_newdoc``.
+add_newdoc_ufunc        It's an internal function and doesn't have a replacement.
+asfarray                Use ``np.asarray`` with a proper dtype instead.
+byte_bounds             Now it's available under ``np.lib.array_utils.byte_bounds``
+cast                    Use ``np.asarray(arr, dtype=dtype)`` instead.
+cfloat                  Use ``np.complex128`` instead.
+clongfloat              Use ``np.clongdouble`` instead.
+compat                  There's no replacement, as Python 2 is no longer supported.
+complex\_               Use ``np.complex128`` instead.
+DataSource              It's still available as ``np.lib.npyio.DataSource``.
+deprecate               Raise ``DeprecationWarning`` instead.
```

Comment:
"raise DeprecationWarning" is not very helpful if one is replacing a decorator that handled that automatically. How about refering to `typing.deprecated` (in python 3.12, backported in `typing_extensions` in earlier versions), since that is much more similar to the old `np.deprecate` that it can be used as a decorator (and is recognized in IDEs). I.e., it is now in the standard library!

Reply:
I think we recognized that this decorator isn't useful / good practice, and have moved away from using it quite a while ago. We've switched to calling `warnings.warn` directly, and prefer that over `typing.deprecated` - the extra complexity and lower readability of using a decorator is not warranted here. So if we'd want to mention the latter, I'd say that it should be after our actual recommendation.

I propose: "emit `DeprecationWarning` with `warnings.warn` directly, or use `typing.deprecated`."